sched/sched/sched_reprioritizertr.c

```diff
@@ -57,34 +57,67 @@ bool nxsched_reprioritize_rtr(FAR struct tcb_s *tcb, int priority)
 {
   bool switch_needed;
 
-  /* Remove the tcb task from the ready-to-run list.
-   * nxsched_remove_readytorun will return true if we just
-   * remove the head of the ready to run list.
+  /* We are not allowed to do a context switch if priority of a currently
+   * running task changes and the scheduler is locked
    */
 
-  switch_needed = nxsched_remove_readytorun(tcb, false);
+  if (tcb == this_task() && tcb->lockcount > 0)
+    {
+      FAR struct tcb_s *ptcb;
 
-  /* Setup up the new task priority */
+      /* Setup up the new task priority */
 
-  tcb->sched_priority = (uint8_t)priority;
+      tcb->sched_priority = (uint8_t)priority;
 
-  /* Return the task to the specified blocked task list.
-   * nxsched_add_readytorun will return true if the task was
-   * added to the new list.  We will need to perform a context
-   * switch only if the EXCLUSIVE or of the two calls is non-zero
-   * (i.e., one and only one the calls changes the head of the
-   * ready-to-run list).
-   */
+      /* Move all tasks with the higher or equal priority to a newly assigned
+       * priority from ready-to-run to pending list
+       */
 
-  switch_needed ^= nxsched_add_readytorun(tcb);
+      for (ptcb = tcb->flink;
```

Comment:
can we abstract a new function for this loop?

Reply:
I will think about that

Reply:
Actually I would like to talk more about this change as this code is generic, but the issue with `sched_lock()` appears only in non-SMP mode because in case of SMP the `sched_lock()` ends up with
```
      /* Move any tasks in the ready-to-run list to the pending task list
       * where they will not be available to run until the scheduler is
       * unlocked and nxsched_merge_pending() is called.
       */

      nxsched_merge_prioritized(&g_readytorun,
                                &g_pendingtasks,
                                TSTATE_TASK_PENDING);
```
But such approach will not work for non-SMP because non-SMP does not have `g_assignedtasks` list. I tried to experiment with similar approach in https://github.com/apache/nuttx/commit/3831dba95ba5ab94f200f70b55eec5d014f12892, but those changes didn't work well while this changes work well.
@xiaoxiang781216 @masayuki2009 I would like to have your opinion on should we focus on `sched_lock`/`sched_unlock` for non-SMP code or going with this generic approach is fine

Reply:
I've found a few issues in ostest and I am going to provide a fix soon. With those changes I can implement a solution with moving tasks to the pending list in `sched_lock()`. I will start an e-mail thread to discuss the best solution for the sched_lock issue.

Reply:
Here it is https://github.com/apache/nuttx-apps/pull/1714

Reply:
I've reworked the PR and localised the changes near to `sched_lock` and `nxsched_merge_pending` for non-SMP case since SMP case does not suffer from the issue.